### ISSUE TITLE
Funkce encode_source_id

### ISSUE THREAD

**Issue by raugustyn** [raugustyn/TB04CUZK001_R2#17]
Pochopil jsem správně parametry  funkce mezi ogc_fid v Data10 a identifikátorem v naší megatabulce takto:

```
CREATE OR REPLACE FUNCTION encode_source_id(
    feature_class string,
    ogc_fid string)
  RETURNS string AS
...
```

???

**Comment by LabyrinthTechnology** [comment]
Odpověď bude dobrým vstupem do #8

**Comment by to-kr** [comment]
Vypocet source_id bych nedelal funkci v DB. Jeho strukturovany obsah muze byt sestaven ruzne. Nechal bych to mimo na urovni pripravy dat popripade importu prostrednictvim OGR, kde je mozne source_id dynamicky sestavit v SQL dotazu.

**Comment by raugustyn** [comment]
Při primární konverzi samozřejmě ano, nicméně pro ladění a testování je potřeba být schopen se dohledat i uvnitř procesů konkrétních geometrií. Tam i zpět. Model bez jasné vazby na původní data je nepoužitelný.

**Comment by to-kr** [comment]
Souhlas, jen k tomu myslim neni potreba zminena funkce explicitne predpokladajici, ze se source_id sklada z featureclass + fid.

**Comment by raugustyn** [comment]
No, tak jasně. Já jen že někde je potřeba začít, klidně se to potom může vyprecizovat.

**Comment by to-kr** [comment]
Pro import v části SW DATA, za kterou jsem odpovědný, potřebu takové funkce nyní nevidím. Přdávám zpět a navrhuji uzavřít.

**Comment by raugustyn** [comment]
@janlangr, @LabyrinthTechnology: chápu to správně že @to-kr má snahu uzavřít celé řešení tak, aby nebylo možno technicky možné vytvořit ověřenou technologii, která je požadavkem projektu? Do toho s Váma kluci nejdu. Nějaký návrh?

**Comment by LabyrinthTechnology** [comment]
Chtěl jsem na Tomášův názor reagovat už včera, ale nedostal jsem se k tomu :-(

Chápu to správně, že se bavíme o funkci, která zajistí "překlad" identifikátorů mezi datovým modelem "data10" a datovým modelem "megatabulka"? Takže pokud mi z jakéhokoliv důvodu nebudou stačit informace v megatabulce, mohu zadat (předpokládám) elements.source_id a vypadne mi feature type (ať již nějaký identifikátor nebo jméno tabulky) a ogc_id z data10 a dostanu se tak na původní data?

Tohle určitě potřebujeme (tam i zpět). Předpokládám 50% funkčnosti už je někde v importu dat (nedíval jsem se, jak přesně import funguje) a druhá půka bude součástí exportu zpět do modelu data10.

Možná je takový překlad triviální, v této chvíli mi ale není úplně jasné, jak se pro daný feature data10 záznam v megatabulce vytvoří. To se předpokádám vyjasní, až bude k dispozici popis datového modelu (řešeno v #8).

**Comment by raugustyn** [comment]
@LabyrinthTechnology: Ano, mluvíme o jedné ze dvou funkcí, které jsme v Hradci označili závorkama aby bylo jasné, že se bude jednat o obecnou funkci sloužící jako interface na vstupu:
<img width="526" alt="mindmap" src="https://cloud.githubusercontent.com/assets/3672235/15699601/77207f60-27ce-11e6-9986-54b83e932ad8.png">

**Comment by raugustyn** [comment]
Poprosil bych kolegu Langra o vyjádření. Díky moc.

**Comment by janlangr** [comment]
Prezentujeme naše řešení jako black box, do kterého vstoupí data ve výchozím DM společně s konfigurací, která stanoví namapování prvků (elementů) mezi vstupním modelem a pracovním modelem black boxu. Tatáž konfigurace zároveň zajistí namapování prvků mezi pracobním modelem a výstupním modelem (ten nemusí být totožný - uvažujeme např. kolaps prvků mezi geometrickými typy). Z pohledu ověřené technologie nevnímám žádný problém v tom, zda a případně jaká funkce tento překlad do megatabulky zajistí. Tomáš pouze uvedl, že z pohledu M1 nevidí potřebu takové interní funkce, neboť všechny potřebné informace byly do megatabulky zaneseny při importu. Ukáže-li se, že nějaký podstatný údaj chybí, pak se předpokládám upraví DM megatabulky a doplní proces importu.

**Comment by raugustyn** [comment]
Chápu, díky.

> všechny potřebné informace byly do megatabulky zaneseny při importu

To platí z úhlu pohledu blackboxu M1. Import ogr2ogr nefunguje spráně, vypařila se geometrie generalizovaných prvků, kterou zatím rekonstruujeme dodatečně. Je to temná stránka OpenSource, tzv. neživý kod.

Řešením asi bude specifický vstup pro M1 přesně podle toho jak jsme se dohodli, to bude jednoduchej unionizovanej dotaz do megatabulky. Na výstupu potom logicky sekvence rozhození nazpátek do obrazu Data10.

Tedy:
source_id = FeatureClassName + ":" + ogc_fid
source_elt_type = znacka

**Comment by janlangr** [comment]
Tím ogc_fid máte na mysli ObjectID vstupní databáze? Je to tak možné, ale spíše jsem předpokládal, že source_id bude vytvořeno z údajů, které jsou v datech Data 10 obsaženy, ale jsou nezávislé na jeho DM, tj. položek source (původ prvku) a source_id (jednoznačný identifikátor prvku v rámci DB původu). Má to však své ALE neboť ne všechny prvky mají source_id (prvky převzaté z MGE).
Aby platilo source_elt_type = znacka, bude muset být na vstupu sedlčanskými provedena určitá konsolidace dat (to plyne z mé dosud nedokončené analýzy Data 10).

**Comment by raugustyn** [comment]
Tomáš navrhuje odchýlit se od zdrojového datového modelu zákazníka, ke kterému se budeme na konci samozřejmě potřebovat vrátit. K tomu se vždy doporučuje jedinečný identifikátor v rámci tabulky, v našem případě ogc_fid, což je pořadové číslo záznamu, počínajíc od jedničky. Tam nejsme vázání na Sedlčanské.

Obdobně source_elt_type, které jsem pochpil tak, že se jedná o interní M1 identifikaci symbolu použitého pro vykreslení.

Asi se vrátíme k původní variantě, kdy každý modul zgeneralizuje obraz data10 v databázi a posune to dál.

**Comment by janlangr** [comment]
Předpokládám, že pokud je ambicí možnost porovnat vstupní data a výstupní data na úrovni jednotlivých prvků, pak potřebujeme převzít, udržet a vrátit do výstupních dat jedinečný identifikátor prvku ve vstupních datech, tedy z mého pohledu nestačí ogc_fid po importu.

Source_elt_type je jednoznačná identifikace kategorie elementu s vlastnostmi společnými pro všechny elementy dané kategorie. Je kategorie je postavena na symbolu definovaného Katalogem mapových značek a datovým modelem. Nedovedu si představit, že tato kategorizace nebude společná pro všechny moduly (viz závěry workshopu).

**Comment by raugustyn** [comment]
> Předpokládám, že pokud je ambicí možnost porovnat vstupní data a výstupní data na úrovni jednotlivých prvků, pak potřebujeme převzít, udržet a vrátit do výstupních dat jedinečný identifikátor prvku ve vstupních datech, 

Ano

> tedy z mého pohledu nestačí ogc_fid po importu

Ne, právě identidikace ogc_fid vede v závěru k možnosti správně aktualizovat záznamy v ArcSDE.

> Source_elt_type je jednoznačná identifikace kategorie elementu s vlastnostmi společnými pro všechny elementy dané kategorie. Je kategorie je postavena na symbolu definovaného Katalogem mapových značek a datovým modelem. Nedovedu si představit, že tato kategorizace nebude společná pro všechny moduly

Zatím jsem to neviděl a čas letí, já si to neumím představit o co jde. Podle mne se jedná o parametrizaci na straně uživatele, kterou si připraví samostatně v Sedlčanech.

> Nedovedu si představit, že tato kategorizace nebude společná pro všechny moduly

Já předpokládám, že toto bude prací Sedlčanských. Pokud vytvoříme něco většího společného, budu rád, ale není to mandatorní. V klidu se omezíme na těch pár prvků z generalizačních situací. Jsem dalek toho slibovat uživateli vzdušné zámky, které potom nepůjde realizovat. Viz například pokus o export všech obrysů značek do PostGIS.

> (viz závěry workshopu).

Závěr workshopu ve směru datového modelu byl, že Tomáš připraví konverzi tak, abysme na tom pochopili přesně o čem mluví a byli schopni konkretizovat vágně definované věci, tj. například právě tyto sloupce o kterých hovořím. Jedná se o návrh.
